File: app/src/main/java/com/example/schoolairdroprefactoredition/scene/user/UserAvatarViewModel.java
package com.example.schoolairdroprefactoredition.scene.user;

import android.app.Application;

import androidx.annotation.NonNull;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;

import com.example.schoolairdroprefactoredition.domain.DomainAvatarUpdateResult;
import com.example.schoolairdroprefactoredition.presenter.callback.IUserAvatarCallback;
import com.example.schoolairdroprefactoredition.presenter.impl.UserAvatarImpl;
import com.example.schoolairdroprefactoredition.scene.main.base.BaseStateAndroidViewModel;

public class UserAvatarViewModel extends BaseStateAndroidViewModel implements IUserAvatarCallback {

    private final MutableLiveData<DomainAvatarUpdateResult> mUpdateCallback = new MutableLiveData<>();

    private final UserAvatarImpl userAvatarImpl;

    public UserAvatarViewModel(@NonNull Application application) {
        super(application);
        userAvatarImpl = UserAvatarImpl.getInstance();
        userAvatarImpl.registerCallback(this);
    }

    public LiveData<DomainAvatarUpdateResult> updateAvatar(String token, String img) {
        userAvatarImpl.updateAvatar(getApplication(), token, img);
        return mUpdateCallback;
    }


    @Override
    public void onUpdateSuccess(DomainAvatarUpdateResult result) {
        mUpdateCallback.postValue(result);
    }

    @Override
    public void onError() {
        mOnRequestListener.onError();
    }


}

File: app/src/main/java/com/example/schoolairdroprefactoredition/viewmodel/UserAvatarViewModel.kt
package com.example.schoolairdroprefactoredition.viewmodel

import android.app.Application
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.viewModelScope
import com.example.schoolairdroprefactoredition.repository.UploadRepository
import com.example.schoolairdroprefactoredition.repository.UserAvatarRepository
import com.example.schoolairdroprefactoredition.utils.ConstantUtil
import kotlinx.coroutines.launch

class UserAvatarViewModel(application: Application) : AndroidViewModel(application) {

    private val userAvatarRepository by lazy {
        UserAvatarRepository.getInstance()
    }

    private val uploadRepository by lazy {
        UploadRepository.getInstance()
    }

    /**
     * 更新用户头像
     */
    fun updateAvatar(token: String?, avatarLocalPath: String): LiveData<String?> {
        val updateLiveData = MutableLiveData<String?>()
        if (token != null) {
            viewModelScope.launch {
                val localPath = ArrayList<String>(1).also { it.add(avatarLocalPath) }
                // 上传图片至七牛云
                uploadRepository.upload(token, localPath, ConstantUtil.UPLOAD_TYPE_AVATAR) { taskAndKeys ->
                    if (taskAndKeys != null) {
                        userAvatarRepository.updateAvatar(token, taskAndKeys.taskId, taskAndKeys.keys[0]) {
                            updateLiveData.postValue(it)
                        }
                    } else {
                        updateLiveData.postValue(null)
                    }
                }
            }
        } else {
            updateLiveData.postValue(null)
        }
        return updateLiveData
    }
}